### PR TITLE
[REVIEW] Use enum class rather than enum for mask_allocation_policy

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Use enum class rather than enum for mask_allocation_policy

Also adding private header for copy

### CODE REVIEW

File: cpp/include/cudf/copying.hpp [hunk 1]
```diff
@@ -19,18 +19,18 @@
...
 * @brief Indicates when to allocate a mask, based on an existing mask.
 * ---------------------------------------------------------------------------**/
-enum mask_allocation_policy {
+enum class  mask_allocation_policy {
     NEVER, ///< Do not allocate a null mask, regardless of input
     RETAIN, ///< Allocate a null mask if the input contains one
     ALWAYS ///< Allocate a null mask, regardless of input
 };
 
+namespace cudf {
```
Comment (trevorsm7) [inline cpp/include/cudf/copying.hpp]: Keep mask_allocation_policy in the cudf::experimental namespace. The rest looks good

**Comment by rgsl888prabhu** [comment]
@harrism @jrhemstad

File: cpp/src/copying/copy_detail.hpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * @param stream Optional CUDA stream on which to execute kernels
+ * @return std::unique_ptr<column> An empty column of same type as `input`
+ */
+std::unique_ptr<column> empty_like(column_view input, cudaStream_t stream = 0);
+
+/**
+ * @brief Creates an uninitialized new column of the same size and type as the `input`.
+ * Supports only fixed-width types.
+ *
+ * @param input Immutable view of input column to emulate
```
Comment (harrism) [inline cpp/src/copying/copy_detail.hpp]: Please document in/out inout with [in] etc.

Comment (harrism) [reply cpp/src/copying/copy_detail.hpp]: Also in the public API.

File: cpp/src/copying/copy.cpp [hunk 3]
```diff
@@ -20,7 +20,7 @@
 #include <cudf/column/column.hpp>
 #include <cudf/table/table.hpp>
 #include <cudf/utilities/traits.hpp>
-#include "copy_detail.hpp"
+#include <detail/copy_detail.hpp>
```
Comment (harrism) [inline cpp/src/copying/copy.cpp]: Should this be <cudf/detail/copy_detail.hpp>?

Comment (rgsl888prabhu) [reply cpp/src/copying/copy.cpp]: As per Jake's guide the path is `cudf/cpp/include/detail/`

https://github.com/rapidsai/cudf/blob/ff1b66c9e3ab260914246a7b78bd2aa42cbdb301/cpp/docs/TRANSITIONGUIDE.md#streams

Comment (harrism) [reply cpp/src/copying/copy.cpp]: Should be `cudf/cpp/include/cudf/detail` for consistency. Jake updated guidance. :)

**Comment by rgsl888prabhu** [comment]
@harrism

File: cpp/include/cudf/detail/copy_detail.hpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
+/*
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/copy_detail.hpp]: rename file to `copy.hpp`.

**Comment by harrism** [comment]
Failing timestamp tests?

**Comment by rgsl888prabhu** [comment]
> Failing timestamp tests?

Yes, after merging with latest branch 0.11, it was working before. Even David Wendt mentioned about it.

**Comment by rgsl888prabhu** [comment]
> After having done minimal research in to the subject, it looks like enum classes should be used wherever possible to reduce confusion (such as `RETAIN` vs `mask_allocation_policy:RETAIN` ). Is that the general guidance?

@jrhemstad might have proper answer for it.

**Comment by jrhemstad** [comment]
> After having done minimal research in to the subject, it looks like enum classes should be used wherever possible to reduce confusion (such as `RETAIN` vs `mask_allocation_policy:RETAIN` ). Is that the general guidance?

Yes, with the caveat that Cython doesn't appear to support enum classes, so they should be avoided in external APIs. https://github.com/cython/cython/issues/1603

Though if you want to try and find a workaround, that would be swell.

**Pull request merged**